optimizely/helpers/condition.py

```diff
@@ -12,115 +12,180 @@
-  ConditionalOperatorTypes.NOT
-]
+class ConditionMatchTypes(object):
+  EXACT = 'exact'
+  EXISTS = 'exists'
+  GREATER_THAN = 'gt'
+  LESS_THAN = 'lt'
+  SUBSTRING = 'substring'
+
 
+class CustomAttributeConditionEvaluator(object):
+  """ Class encapsulating methods to be used in audience leaf condition evaluation. """
 
-class ConditionEvaluator(object):
-  """ Class encapsulating methods to be used in audience condition evaluation. """
+  CUSTOM_ATTRIBUTE_CONDITION_TYPE = 'custom_attribute'
 
   def __init__(self, condition_data, attributes):
     self.condition_data = condition_data
-    self.attributes = attributes
+    self.attributes = attributes or {}
 
-  def evaluator(self, condition):
-    """ Method to compare single audience condition against provided user data i.e. attributes.
+  def is_value_valid_for_exact_conditions(self, value):
+    """ Method to validate if the value is valid for exact match type evaluation
 
     Args:
-      condition: Integer representing the index of condition_data that needs to be used for comparison.
+      value: Value to validate
 
     Returns:
-      Boolean indicating the result of comparing the condition value against the user attributes.
+      Boolean: True if value is a string type, or a boolean, or is finite. Otherwise False
     """
+    if isinstance(value, string_types) or isinstance(value, bool) or is_finite_number(value):
+      return True
 
-    return self.attributes.get(self.condition_data[condition][0]) == self.condition_data[condition][1]
+    return False
 
-  def and_evaluator(self, conditions):
-    """ Evaluates a list of conditions as if the evaluator had been applied
-    to each entry and the results AND-ed together
+  def exact_evaluator(self, index):
+    """ Evaluate the given exact match condition for the user attributes
 
     Args:
-      conditions: List of conditions ex: [operand_1, operand_2]
+      index: Index of the condition to be evaluated
 
     Returns:
-      Boolean: True if all operands evaluate to True
+      Boolean:
+        - True if the user attribute value is equal (===) to the condition value
+        - False if the user attribute value is not equal (!==) to the condition value
+      None:
+        - if the condition value or user attribute value has an invalid type
+        - if there is a mismatch between the user attribute type and the condition value type
+    """
+    condition_value = self.condition_data[index][1]
+    user_value = self.attributes.get(self.condition_data[index][0])
+
+    if not self.is_value_valid_for_exact_conditions(condition_value) or \
+       not self.is_value_valid_for_exact_conditions(user_value) or \
+       not are_values_same_type(condition_value, user_value):
+        return None
+
+    return condition_value == user_value
+
+  def exists_evaluator(self, index):
+    """ Evaluate the given exists match condition for the user attributes
+
+      Args:
+        index: Index of the condition to be evaluated
```

Comment:
Same comment as above. Please update all instances of index in this PR to be something different.

Reply:
I think this is actually clean after all!